File: Programmers/Lv3/Lv3_섬연결하기.py
# ?꾨┝?뚭퀬由ъ쬁(???④퀎?먯꽌 留뚮뱾?댁쭊 ?좎옣 ?몃━ 吏묓빀???몄젒???뺤젏??以묒뿉??理쒖냼 媛꾩꽑?쇰줈 ?곌껐???뺤젏???좏깮?섏뿬 ?몃━瑜??뺤옣)
import operator

def isEnd(check, n):
    for i in range(n):
        if check[i] == False:
            return False
    return True

def solution(n, costs):
    answer = 0
    check = [False]*101

    check[costs[0][0]] = True       # ?쒖옉吏??True

    while not isEnd(check,n):
        island = []
        for i in range(n):
            if check[i]:        # 寃?ы븷 ?뺤젏
                for j in costs:
                    if (j[0] == i and check[j[1]] == False) or (j[1] == i and check[j[0]] == False):    # ?꾩쭅 ???ъ슜??媛꾩꽑????
                        island.append(j)

        island.sort(key=operator.itemgetter(2))
        temp = island[0]
        answer += temp[2]
        check[temp[0]] = True
        check[temp[1]] = True
        costs.remove(temp)

    return answer

print(solution(4, [[0, 1, 1], [0, 2, 2], [1, 2, 5], [1, 3, 1], [2, 3, 8]]))     # 4

File: Programmers/Lv3/Lv3_섬연결하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <list>
#include <iostream>
#include <algorithm>
using namespace std;

// 프림 알고리즘 사용
struct cmp_cost {				// top에 가장 작은 수가 가도록 내림차순 정렬
	bool operator()(vector<int> t, vector<int> u) {
		return t[2] > u[2];
	}
};

bool isEnd(bool check[], int n) {
	for (int i = 0; i < n; i++) {
		if (check[i] == false)
			return false;
	}
	return true;
}

int solution(int n, vector<vector<int>> costs) {
	int answer = 0;
	bool check[101] = { false, };
	int pos = 0;
	list<vector<int>> l;

	for (int i = 0; i < costs.size(); i++)
		l.push_back(costs[i]);
		
	check[(*l.begin())[0]] = true;		// 시작시점 설정
	
	while (!isEnd(check, n)) {
		priority_queue <vector<int>, vector<vector<int>>, cmp_cost > pq_island;
		for (int i = 0; i < n; i++) {
			if (check[i]) {		// 추가된 정점
				for (auto j : l) {		// 검사할 간선을 우선순위큐에 삽입
					if ((j[0] == i && check[j[1]]!=true) || (j[1] == i && check[j[0]]!=true))
						pq_island.push(j);
				}
			}
		}
		vector<int> v = pq_island.top();				// 비용이 가장 적은 간선을 선택
		answer += v[2];									// 비용에 추가
		check[v[0]] = true;								// 정점 추가
		check[v[1]] = true;								// 정점 추가
		l.erase(find(l.begin(), l.end(), v));
	}

	return answer;
}